p2p/net/swarm/swarm.go

```diff
@@ -588,9 +607,17 @@ func (s *Swarm) removeConn(c *Conn) {
 				cs[len(cs)-1] = nil
 				s.conns.m[p] = cs[:len(cs)-1]
 			}
-			return
+			break
 		}
 	}
+	s.conns.Unlock()
+
+	if disconnected {
+		s.emitter.Emit(event.EvtPeerConnectednessChanged{
```

Comment:
What's the reason to do this here instead of near the old place where this happened: https://github.com/libp2p/go-libp2p/blob/marco/1575/p2p/net/swarm/swarm_conn.go#L85 ?

Reply:
Probably to ensure that we only emit the event on the last disconnect. But it does look like there's a bit of a race and we can emit a disconnect after a connect (looks like we already _have_ that race, actually).